### PR TITLE
[VSUP-126] Gate SSL certificate validation — only allow self-signed in DEBUG for dev hosts

### PR DESCRIPTION
**Pull request by g-telnyx**
## Summary

Fixes VSDK-423 / Resolves VSUP-126

The SDK was unconditionally disabling SSL certificate validation on both the signaling socket and `CallReportCollector`, allowing MITM attacks on production traffic.

## Changes

- **SSLValidationHelper.swift** (new) — Restricts self-signed cert acceptance to localhost and private/loopback IP ranges only. Prevents hostname spoofing via proper IPv4 literal parsing.
- **Socket.swift** — Gates self-signed cert acceptance behind `#if DEBUG` + `SSLValidationHelper`. Release builds always enforce strict validation.
- **TelnyxCallReportCollector.swift** — `AllowSelfSignedDelegate` gated behind `#if DEBUG` and restricted to local hosts. Release builds use `URLSession.shared` with standard system validation.
- **Tests** — 3 new test files covering local host acceptance, non-local rejection, and hostname spoofing prevention.

## Test Plan

- [ ] CI passes (build + tests)
- [ ] Verify DEBUG build connects to localhost with self-signed cert
- [ ] Verify DEBUG build rejects self-signed cert for non-local hosts
- [ ] Verify RELEASE build enforces strict validation for all hosts

### CODE REVIEW

File: TelnyxRTC/Telnyx/Security/SSLValidationHelper.swift [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        if isPrivateOrLoopbackIP(host) {
+            return true
+        }
+
+        return false
+    }
+
+    private static func isPrivateOrLoopbackIP(_ host: String) -> Bool {
```
Comment (gbattistel) [inline TelnyxRTC/Telnyx/Security/SSLValidationHelper.swift]: ⚠️ **IPv6 not supported.** `isPrivateOrLoopbackIP` only handles 4-octet IPv4, so `::1` and ULA `fc00::/7` are not recognized. Denying is always safe, but a dev server on `::1` won't get the DEBUG bypass. Consider adding IPv6 loopback/ULA checks (or note the limitation explicitly).

File: TelnyxRTC/Telnyx/Services/Socket.swift [hunk 2]
```diff
@@ -47,7 +47,18 @@ class Socket {
         
         var request = URLRequest(url: signalingServer)
         request.timeoutInterval = connectionTimeout
-        let pinner = FoundationSecurity(allowSelfSigned: true) // don't validate SSL certificates
+
+        let pinner: FoundationSecurity
+        #if DEBUG
```
Comment (gbattistel) [inline TelnyxRTC/Telnyx/Services/Socket.swift:52]: ⚠️ This gating is correct, but it only matches `localhost`/loopback/private IPv4 **literals**. If any dev/staging environment is reached via a **hostname** that resolves to a private IP, DEBUG builds can no longer accept its self-signed cert and connections will fail. Please confirm with the Voice team that all dev endpoints use localhost/IP literals, or add a hostname allowlist / DNS resolution step.

File: TelnyxRTC/Telnyx/WebRTC/Stats/TelnyxCallReportCollector.swift [hunk 3]
```diff
@@ -262,29 +262,46 @@ public class TelnyxCallReportCollector {
...
             } else {
                 completionHandler(.performDefaultHandling, nil)
             }
         }
     }
+    #endif
 
-    private lazy var urlSession: URLSession = {
+    #if DEBUG
+    private lazy var localSession: URLSession = {
```
Comment (gbattistel) [inline TelnyxRTC/Telnyx/WebRTC/Stats/TelnyxCallReportCollector.swift]: 💡 `lazy var` on a class is not thread-safe under concurrent first access — two simultaneous `executeUpload` calls could initialize it twice. Current usage appears serial, but initializing eager or using a `static let` would be safer.

File: TelnyxRTC/Telnyx/WebRTC/Stats/TelnyxCallReportCollector.swift [hunk 4]
```diff
@@ -262,29 +262,46 @@ public class TelnyxCallReportCollector {
...
+        }
+        return URLSession.shared
+        #else
+        return URLSession.shared
+        #endif
+    }
 
     private func executeUpload(request: URLRequest, attempt: Int) {
-        let task = urlSession.dataTask(with: request) { [weak self] data, response, error in
+        let session = urlSession(for: request.url!)
```
Comment (gbattistel) [inline TelnyxRTC/Telnyx/WebRTC/Stats/TelnyxCallReportCollector.swift]: ⚠️ `request.url!` force-unwrap. `URLRequest(url:)` always sets a non-nil URL in practice, but the unwrap is fragile. Prefer `guard let url = request.url else { return }`.

File: TelnyxRTCTests/WebRTC/CallReportCollectorSSLTests.swift [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    func testCallReportHelperAllowsLocalhost() {
+        let url = URL(string: "wss://localhost")!
+        XCTAssertTrue(SSLValidationHelper.shouldAllowSelfSigned(for: url))
+    }
+
+    // MARK: - Post report
+
+    func testPostReportDoesNotCrash() {
```
Comment (gbattistel) [inline TelnyxRTCTests/WebRTC/CallReportCollectorSSLTests.swift]: 💡 `testPostReportDoesNotCrash` posts to `wss://rtc.telnyx.com` (non-local) but only asserts no crash — it never verifies that `URLSession.shared` (strict validation) was selected. The critical "non-local → shared session" path is untested functionally. Consider injecting the session or asserting the chosen session type.

File: TelnyxRTCTests/Security/SocketSSLConfigurationTests.swift [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        let socket = Socket()
+        socket.automaticallyConnectWebSocket = false
+
+        let url = URL(string: "wss://localhost")!
+        XCTAssertTrue(SSLValidationHelper.shouldAllowSelfSigned(for: url))
+    }
+
+    // MARK: - API surface
+
+    func testSSLValidationHelperHasStaticMethod() {
```
Comment (gbattistel) [inline TelnyxRTCTests/Security/SocketSSLConfigurationTests.swift]: 💡 `testSSLValidationHelperHasStaticMethod` and `testDevHostsContainsLocalhost` (line 62) only assert the method exists and that a string is in a list — they add no real coverage and could be removed or strengthened.

**Review by gbattistel — CHANGES_REQUESTED** [review]
## Code Review — Hermes Agent

> **Update:** The version bump / CHANGELOG point raised in the first version of this review is **not applicable** — the CHANGELOG is produced at release time as a separate process, and consumers pick up the SDK via a tag that has not been published yet (neither SPM nor CocoaPods). That section has been removed.

**Verdict: Changes Requested** — for a dev-environment functional risk and a few minor issues. The security fix itself is sound.

Gating `allowSelfSigned` behind `#if DEBUG` + `SSLValidationHelper` (localhost/loopback/private IPv4 only) and using `FoundationSecurity(allowSelfSigned: false)` in release is the right approach. IPv4 literal parsing with `UInt8(octetStr)` correctly prevents hostname spoofing, and the boundary tests for `172.15`/`172.32` are good.

### ⚠️ Warnings

- **Dev/staging hosts that use a domain (not a private IP literal) lose the DEBUG bypass.** `SSLValidationHelper` only matches `localhost`, `127.x`, `10.x`, `192.168.x`, `172.16-31.x`. If any dev/staging endpoint is a hostname that *resolves* to a private IP (e.g. `dev.voice.telnyx.com`), DEBUG builds can no longer connect to a self-signed cert there and will fail. Please confirm with the Voice team that all dev environments use localhost/IP literals — otherwise add a hostname allowlist or resolve the DNS before deciding.

- **IPv6 not supported.** `isPrivateOrLoopbackIP` only parses 4 IPv4 octets, so `::1` and ULA `fc00::/7` are not recognized. This is never a security problem (deny is always safe), but a dev server bound to `::1` won't get the bypass in DEBUG.

- **`request.url!` force-unwrap** — see inline.

### 💡 Suggestions

- Tests like `testSSLValidationHelperHasStaticMethod` and `testDevHostsContainsLocalhost` only assert the method exists / a string is in a list — they add no real coverage.
- `testPostReportDoesNotCrash` posts to `wss://rtc.telnyx.com` but only checks it doesn't crash; it never verifies *which* `URLSession` was selected. The critical "non-local → `URLSession.shared`" path is untested functionally.
- The test target compiles in DEBUG, so the `#else` (RELEASE) branches of `urlSession(for:)` and `Socket.connect` are never exercised in CI. Consider a contract test under `#if !DEBUG` or document the limitation.
- `lazy var localSession` is not thread-safe under concurrent first access (Swift `lazy` on classes). Current usage looks serial, but initializing eager or using a `static let` would be safer.

### ✅ Looks Good

- `#if DEBUG` excludes the insecure code path from the release binary entirely.
- Correct private-range logic with boundary tests.
- `project.pbxproj` properly updated (file refs, groups, build phases).

---
*Reviewed by Hermes Agent*

**Comment by g-telnyx** [comment]
## Question on dev/staging hostname support

The review flagged that `SSLValidationHelper` only matches `localhost`, loopback, and private IPv4 literals — so dev/staging environments accessed via a **hostname** (e.g. `dev.voice.telnyx.com`) that resolves to a private IP won't get the DEBUG self-signed bypass.

Two options:

1. **Add a hostname allowlist** — hardcode known dev/staging hostnames in the SDK so the helper can match them directly.
2. **DNS resolution** — resolve the hostname at runtime and check if the resolved IP is private/loopback.

Option 1 is simpler and safer, but it **leaks dev/staging hostnames into the SDK binary**. Even though it's `#if DEBUG`-gated, the hostnames would be visible in the compiled framework.

**Question for the team:** Are we comfortable exposing dev/staging hostnames in the SDK (DEBUG-only), or should we keep the current IP-literal-only approach and require dev environments to use IP literals / localhost?

If all dev/staging environments already use localhost or IP literals, the current code is fine as-is and we just need to document the limitation.

File: TelnyxRTC/Telnyx/Security/SSLValidationHelper.swift (hunk 1 above)
Comment (g-telnyx) [reply TelnyxRTC/Telnyx/Security/SSLValidationHelper.swift]: Fixed. `SSLValidationHelper` now supports:
- IPv6 loopback (`::1`)
- IPv6 ULA (`fc00::/7` — `fc` and `fd` prefixes)
- IPv6 link-local (`fe80::/10`)
- IPv4-in-IPv6 mapping (`::ffff:127.0.0.1`)

Tests added: `testAllowsIPv6Loopback`, `testAllowsIPv6UniqueLocalFC`, `testAllowsIPv6UniqueLocalFD`, `testRejectsIPv6Public`, `testAllowsIPv4MappedIPv6Loopback`, `testAllowsIPv4MappedIPv6Private`, `testRejectsIPv4MappedIPv6Public`.

File: TelnyxRTCTests/WebRTC/CallReportCollectorSSLTests.swift (hunk 5 above)
Comment (g-telnyx) [reply TelnyxRTCTests/WebRTC/CallReportCollectorSSLTests.swift]: Strengthened. Added `testPostReportToLocalHostDoesNotCrash` which exercises the localhost → `localSession` (self-signed) path, complementing the existing non-local test. Both paths are now covered. Note: full session-type injection would require refactoring `urlSession(for:)` to be injectable — left as a follow-up since the current tests verify both code paths execute without crashing.

File: TelnyxRTC/Telnyx/Services/Socket.swift (hunk 2 above)
Comment (g-telnyx) [reply TelnyxRTC/Telnyx/Services/Socket.swift:52]: Posted a separate comment on the PR asking the team whether dev/staging hostnames should be leaked into the SDK (DEBUG-only) or if IP-literal-only is the intended approach. Keeping the current IP-literal-only behavior until we get confirmation. If hostname support is needed, we can add an allowlist or DNS resolution step in a follow-up.

File: TelnyxRTC/Telnyx/WebRTC/Stats/TelnyxCallReportCollector.swift (hunk 3 above)
Comment (g-telnyx) [reply TelnyxRTC/Telnyx/WebRTC/Stats/TelnyxCallReportCollector.swift]: Fixed. Replaced `lazy var localSession` with `private static let localSession` for thread-safe initialization. Swift guarantees `static let` is initialized lazily and atomically.

File: TelnyxRTC/Telnyx/WebRTC/Stats/TelnyxCallReportCollector.swift (hunk 4 above)
Comment (g-telnyx) [reply TelnyxRTC/Telnyx/WebRTC/Stats/TelnyxCallReportCollector.swift]: Fixed. Replaced `request.url!` with `guard let requestUrl = request.url else { ... return }` for safe unwrapping.

File: TelnyxRTCTests/Security/SocketSSLConfigurationTests.swift (hunk 6 above)
Comment (g-telnyx) [reply TelnyxRTCTests/Security/SocketSSLConfigurationTests.swift]: Fixed. Removed `testSSLValidationHelperHasStaticMethod` and `testDevHostsContainsLocalhost`. Replaced with 23 substantive tests covering IPv4 boundaries (172.15/172.32 reject), IPv6 loopback/ULA, IPv4-mapped IPv6, invalid IPs, and edge cases (empty host, invalid octets).

File: TelnyxRTC.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -1559,7 +1591,7 @@
 					"$(BUILT_PRODUCTS_DIR)/Starscream",
 				);
 				MARKETING_VERSION = 4.0.2;
-				PRODUCT_BUNDLE_IDENTIFIER = com.telnyx.WebRTCSDK;
+				PRODUCT_BUNDLE_IDENTIFIER = com.telnyx.WebRTCSDK.gp;
```
Comment (isaacakakpo1) [inline TelnyxRTC.xcodeproj/project.pbxproj]: Looks like local Xcode signing changes slipped in here. The SDK bundle identifier should stay as `com.telnyx.WebRTCSDK`; the `.gp` suffix and related `DEVELOPMENT_TEAM` / demo/test bundle ID changes look personal to the local dev environment. Can we revert those project settings so this PR only contains the SSL validation changes?

File: TelnyxRTC/Telnyx/Security/SSLValidationHelper.swift [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        // fc00::/7 → unique-local addresses (fc00:: – fdff::)
+        if host.hasPrefix("fc") || host.hasPrefix("fd") {
+            // Validate it's a valid IPv6 address by attempting to parse components
+            let fullForm = expandIPv6(host)
+            guard !fullForm.isEmpty else { return false }
+            return true
+        }
+
+        // fe80::/10 → link-local (also private, but typically not used for dev servers)
+        if host.hasPrefix("fe80") {
```
Comment (isaacakakpo1) [inline TelnyxRTC/Telnyx/Security/SSLValidationHelper.swift]: This should validate that `host` is an actual IPv6 literal before allowing the self-signed bypass. As written, any hostname starting with `fe80` would pass here in DEBUG, e.g. `fe80.attacker.example`, which reintroduces a hostname-spoofing gap. Could we parse the host as IPv6 first, for example with `inet_pton(AF_INET6, host, &addr) == 1`, and only then apply the `::1` / `fc00::/7` / `fe80::/10` checks?

File: TelnyxRTCTests/WebRTC/CallReportCollectorSSLTests.swift [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+    func testHTTPURLNonLocalRejected() {
+        let httpUrl = URL(string: "https://rtc.telnyx.com/report")!
+        XCTAssertFalse(SSLValidationHelper.shouldAllowSelfSigned(for: httpUrl))
+    }
+
+    // MARK: - IPv6 support for call report endpoints
+
+    func testIPv6LoopbackAllowedForCallReport() {
+        let url = URL(string: "wss://[::1]:8080")!
```
Comment (isaacakakpo1) [inline TelnyxRTCTests/WebRTC/CallReportCollectorSSLTests.swift:138]: This confirms the helper allows an IPv6 localhost URL, but it does not cover the actual call-report endpoint builder. That path currently interpolates `wsUrl.host` directly, so an IPv6 literal would become something like `https://::1:8080/call_report` instead of the valid bracketed form `https://[::1]:8080/call_report`. Since this PR adds IPv6 localhost support, can we also cover/fix that endpoint reconstruction path?

File: TelnyxRTC/Telnyx/WebRTC/Stats/TelnyxCallReportCollector.swift [hunk 10]
```diff
@@ -262,29 +262,50 @@ public class TelnyxCallReportCollector {
...
-    ///
-    /// ⚠️ Security Note: This accepts ALL self-signed certificates for dev/staging parity.
-    /// In production, this enables MITM attacks on call reports. Consider production-gating
-    /// this behavior or restricting to known dev/staging hostnames only.
+    #if DEBUG
     private class AllowSelfSignedDelegate: NSObject, URLSessionDelegate {
         func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge,
                         completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
+            let host = challenge.protectionSpace.host
+            guard let checkUrl = URL(string: "https://\(host)"),
```
Comment (isaacakakpo1) [inline TelnyxRTC/Telnyx/WebRTC/Stats/TelnyxCallReportCollector.swift]: Same IPv6 issue here: when `protectionSpace.host == "::1"`, this builds `https://::1`, which is invalid. The guard falls back to default handling, so the DEBUG self-signed bypass won’t actually apply for IPv6 localhost call-report uploads. We should bracket IPv6 literals before creating the URL, or avoid URL reconstruction and validate the host directly.

**Review by isaacakakpo1 — CHANGES_REQUESTED** [review]
Thanks for the update — the original review items are mostly addressed, and the DEBUG/Release SSL split is much stronger now. I’m still requesting changes for two merge blockers:

1. Local Xcode signing/bundle identifier changes appear to have slipped into `project.pbxproj`.
2. The new IPv6 handling should parse actual IPv6 literals before allowing DEBUG self-signed bypass, and the call-report URL reconstruction needs to handle bracketed IPv6 hosts.

Once those are cleaned up, this should be in good shape.

**Review by gbattistel — APPROVED** [review]
All review items addressed — IPv6 inet_pton validation, bracketed IPv6 URLs, local signing reverted, stronger tests. LGTM.

**Review by isaacakakpo1 — APPROVED** [review]
Looks Good. Thanks
